drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+
+		addr = data->address & I2C_SLAVE_ADDR_V;
+		*config->regs.slave_addr |= addr << I2C_SLAVE_ADDR_S;
+		if (dev_config.bits.use_10_bit_addr) {
+			*config->regs.slave_addr |= I2C_ADDR_10BIT_EN;
+		}
+
+		/* Before setting up FIFO and interrupts, stop transmission */
+		*config->regs.ctr &= ~I2C_TRANS_START;
+
+		/* Byte after address isn't the offset address in slave RAM */
+		*config->regs.fifo_cfg &= ~I2C_FIFO_ADDR_CFG_EN;
+	}
+
+	/* Use open-drain for clock and data pins */
+	v |= (I2C_SCL_FORCE_OUT | I2C_SDA_FORCE_OUT);
+
+	v |= I2C_CLK_EN;
+
+	/* Sample data when SCL is high */
+	v &= ~I2C_SAMPLE_SCL_LEVEL;
+
+	*config->regs.ctr = v;
+
+	switch (dev_config.bits.speed) {
+	case I2C_SPEED_STANDARD:
+		i2c_esp32_configure_frequency(config, 100000);
+		break;
+	case I2C_SPEED_FAST:
+		i2c_esp32_configure_frequency(config, 400000);
+		break;
+	default:
+		r = -EINVAL;
+		goto out;
+	}
+
+	/* Use FIFO to transmit data */
+	*config->regs.fifo_cfg &= ~I2C_NONFIFO_EN;
+
+	v = CONFIG_I2C_ESP32_TIMEOUT & I2C_TIME_OUT_REG;
+	*config->regs.timeout = v << I2C_TIME_OUT_REG_S;
+
+	/* Enable interrupt types handled by the ISR */
+	*config->regs.intr |= I2C_ACK_ERR_INT_ENA_M |
+			      I2C_TIME_OUT_INT_ENA_M |
+			      I2C_TRANS_COMPLETE_INT_ENA_M |
+			      I2C_ARBITRATION_LOST_INT_ENA_M;
+
+	/* Activate the device IRQ */
+	irq_enable(config->irq.line);
+
+out:
+	irq_unlock(key);
+
+	return r;
+}
+
+static inline void i2c_esp32_ack_intrs(const struct i2c_esp32_config *config)
+{
+	/* Acknowledge all interrupts */
+	*config->regs.intr_clr = ~0;
+}
+
+
+static inline void i2c_esp32_reset_fifo(const struct i2c_esp32_config *config)
+{
+	/* Writing 1 and then 0 to these bits will reset the I2C fifo */
+	*config->regs.fifo_cfg |= (I2C_TX_FIFO_RST | I2C_RX_FIFO_RST);
+	*config->regs.fifo_cfg &= ~(I2C_TX_FIFO_RST | I2C_RX_FIFO_RST);
+}
+
+static inline void i2c_esp32_prepare(const struct i2c_esp32_config *config)
+{
+	i2c_esp32_ack_intrs(config);
+	i2c_esp32_reset_fifo(config);
+}
+
+static int i2c_esp32_spin_yield(int *counter)
+{
+	*counter = *counter + 1;
+
+	if (*counter > I2C_ESP32_TIMEOUT) {
+		return -ETIMEDOUT;
+	}
+
+	if (*counter > I2C_ESP32_SPIN_THRESHOLD) {
+		k_yield();
+	}
+
+	return 0;
+}
+
+static int i2c_esp32_transmit_and_wait(struct device *dev,
+				       volatile struct i2c_esp32_cmd *wait_cmd)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	struct i2c_esp32_data *data = dev->driver_data;
+	u32_t status;
+	int counter = 0;
+	int r;
+
+	i2c_esp32_ack_intrs(config);
+
+	/* Start transmission and wait for the ISR to give the semaphore */
+	*config->regs.ctr |= I2C_TRANS_START;
+	if (k_sem_take(&data->fifo_sem, I2C_ESP32_TIMEOUT_MS) < 0) {
+		return -ETIMEDOUT;
+	}
+
+	/* Check for error conditions before waiting for the bus to finish */
+	status = *config->regs.intr_raw;
+	if (status & (I2C_ARBITRATION_LOST_INT_RAW | I2C_ACK_ERR_INT_RAW)) {
+		return -EIO;
+	}
+	if (status & I2C_TIME_OUT_INT_RAW) {
```

Comment:
add an empty line here

Reply:
That would hurt readability: spaces are used to group things. In this case, this is grouping the reading of status bits. Adding a space would remove this visual clue.

Reply:
@tbursztyka would say: "Then use "else if" [and rely on compiler to not produce superfluous code, which is simple for if-else-return]". But I'm personally all for letting a programmer not to put a blank line where they don't want ;-).

Reply:
please @pfalcon stop assuming what people would say.